tsdb/shard.go

```diff
@@ -1870,19 +1929,125 @@ func (fs *MeasurementFieldSet) writeToFile(first writeRequest) {
 			return true, err
 		}
 
-		// Read all the pending new field and measurement write requests
-		// that will be captured in the marshaling of the in-memory copy
+		// Lock, copy, and marshal the in-memory index
+		b, err := fs.marshalMeasurementFieldSetNoLock()
+		if err != nil {
+			return true, err
+		}
+		if b == nil {
+			// No fields, file removed, all done
+			return true, nil
+		}
+		if _, err := fd.Write(b); err != nil {
+			return true, err
+		}
+		return false, fd.Sync()
+	}()
+	if err != nil {
+		return err
+	} else if isEmpty {
+		// remove empty file
+		if err = os.RemoveAll(fs.path); err != nil {
+			return err
+		} else {
+			return os.RemoveAll(fs.fieldChangesDirectory())
+		}
+	}
+	if err = fs.renameFile(path); err != nil {
+		return err
+	}
+	return os.RemoveAll(fs.fieldChangesDirectory())
+}
+
+func (fs *MeasurementFieldSet) getNextChangeFile() (fName string, rErr error) {
+	d := fs.fieldChangesDirectory()
+	names, err := filepath.Glob(path.Join(d, changeFilePattern))
+	if err != nil {
+		return "", fmt.Errorf("failed looking for fields change files: %w", err)
+	}
+
+	maxIdxl := 0
+	for _, fn := range names {
+		var curIdxl int
+		if curIdxl, err = changeIndex(fn); err != nil {
+			return "", err
+		}
+		if curIdxl > maxIdxl {
+			maxIdxl = curIdxl
+		}
+	}
+	return path.Join(d, fmt.Sprintf(changeFileFormat, 1+maxIdxl)), nil
+}
+
+func changeIndex(changeFile string) (int, error) {
+	var curIdxl int
+	if m, err := fmt.Sscanf(filepath.Base(changeFile), changeFileFormat, &curIdxl); err != nil {
+		return -1, err
+	} else if m != 1 {
+		return -1, fmt.Errorf("%q: %w", changeFile, ErrBadFieldChangeFile)
+	} else {
+		return curIdxl, nil
+	}
+}
+
+func (fs *MeasurementFieldSet) fieldChangesDirectory() string {
+	return path.Join(path.Dir(fs.path), changeFileDirectory)
+}
+
+// writeChangesToFile: Write the new index to a temp file and rename when it's sync'd
+func (fs *MeasurementFieldSet) writeChangesToFile(first writeRequest) {
```

Comment:
I thought we discussed writing to a log file instead of this approach that will create tons of files even for normal users with hundreds of gradually-added fields?

I assumed we would re-use or share code with the tsl files, similar to:

```
// appendEntry adds a log entry to the end of the file.
func (f *LogFile) appendEntry(e *LogEntry) error {
	// Marshal entry to the local buffer.
	f.buf = appendLogEntry(f.buf[:0], e)

	// Save the size of the record.
	e.Size = len(f.buf)

	// Write record to file.
	n, err := f.w.Write(f.buf)
	if err != nil {
		// Move position backwards over partial entry.
		// Log should be reopened if seeking cannot be completed.
		if n > 0 {
			f.w.Reset(f.file)
			if _, err := f.file.Seek(int64(-n), io.SeekCurrent); err != nil {
				f.Close()
			}
		}
		return err
	}

	// Update in-memory file size & modification time.
	f.size += int64(n)
	f.modTime = time.Now()

	return nil
}
```

Reply:
The `appendEntry` / `execEntry` from TSL is a nice pattern.

Reply:
@lesam - I thought you were against a single log file, because it was a single bottleneck and point of failure, but I can't find our slack discussion to verify my memory.  Can you send me that conversation? Multiple files have at least two advantages:
- no bottleneck or locking as on writing to a single log file
- later failures (like out of disk space) don't  affect earlier changes
- I had thought you said write-once files like TSM files were the preferable route; these match that goal.

The files are never read in normal operation, and are cleaned up on system shutdown. They live in their own directory, which disappears unless the system crashes.

Reply:
I don't think log files would be any worse than multiple files when disk space is exhausted.

How many extra files would we be talking about for multiple files, for example with that one customer where the fields.idx was a major bottleneck? Lots of files in a directory can itself be a bottleneck. A few hundred files and I'm not worried. A few thousand files, probably OK. Tens of thousands? That's getting into the range where I've seen really bad performance scanning the files in the directory.

Reply:
@gwossum - The highest request from a customer was 150,000 fields.  Normal usage is under 1000 fields.  One thing to note is that this PR combines field creation from multiple write requests. Worst case is one file per new field, if only one field is ever created per write request. But, this code batches all new fields per write request, and also batches write requests that come in close enough in time, as well.

Reply:
@lesam - here's where I thought you discouraged me from using a log with appendation:

https://github.com/influxdata/influxdb/issues/23653#issuecomment-1224664104
> The WAL (or any delta log, like what you're suggesting here) is a lot more prone to issues as we need to update the file, and recover from any possible errors with those updates, instead of a process like TSM files where every file can be written once immutably to a tmp file and then we just have to deal with os.Rename success/failure.

Reply:
From the issue discussion:

> The monitoring Go proc aggregates all the changes on the channel into one change file. This will be at least the changes from one write or delete, and possibly more. depending on load.

I thought you meant we were aggregating all the changes into one file, which I agreed to: https://github.com/influxdata/influxdb/issues/23653#issuecomment-1227701921

We discussed that even with normal usage, it would be easy to get 1000's of files per shard with the PR's approach of a separate file each time we do a field write (I know we batch them, but if the fields don't batch unless they come in at almost the same time.

Reply:
https://github.com/influxdata/influxdb/issues/23653#issuecomment-1228887678 - which is why I suggested `fields.idxl` as the (single) log file name...